crates/bevy_pbr/src/render/shadow_sampling.wgsl

```diff
@@ -126,3 +167,166 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
     return 0.0;
 #endif
 }
+
+// NOTE: Due to the non-uniform control flow in `shadows::fetch_point_shadow`,
+// we must use the Level variant of textureSampleCompare to avoid undefined
+// behavior due to some of the fragments in a quad (2x2 fragments) being
+// processed not being sampled, and this messing with mip-mapping functionality.
+// The shadow maps have no mipmaps so Level just samples from LOD 0.
+fn sample_shadow_cubemap_hardware(light_local: vec3<f32>, depth: f32, light_id: u32) -> f32 {
+#ifdef NO_CUBE_ARRAY_TEXTURES_SUPPORT
+    return textureSampleCompare(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, depth);
+#else
+    return textureSampleCompareLevel(view_bindings::point_shadow_textures, view_bindings::point_shadow_textures_sampler, light_local, i32(light_id), depth);
+#endif
+}
+
+fn tap_shadow_cubemap(
+    position: vec2<f32>,
+    coeff: f32,
+    x_basis: vec3<f32>,
+    y_basis: vec3<f32>,
+    light_local: vec3<f32>,
+    depth: f32,
+    light_id: u32,
+) -> f32 {
+    return sample_shadow_cubemap_hardware(
+        light_local + position.x * x_basis + position.y * y_basis,
+        depth,
+        light_id
+    ) * coeff;
+}
+
+// This more or less does what Castano13 does, but in 3D space. Castano13 is
+// essentially an optimized 2D Gaussian filter that takes advantage of the
+// bilinear filtering hardware to reduce the number of taps needed. This trick
+// doesn't apply to cubemaps, so we manually apply a Gaussian filter over the
+// standard 8xMSAA pattern instead.
+fn sample_shadow_cubemap_gaussian(
+    light_local: vec3<f32>,
+    depth: f32,
+    texel_size: f32,
+    distance_to_light: f32,
+    light_id: u32,
+) -> f32 {
+    // Create an orthonormal basis via Gram-Schmidt normalization so we can
```

Comment:
This isn’t Gram-Schmidt I think. That involves projection and subtraction if I remember correctly.

Reply:
It's equivalent to Gram-Schmidt. https://math.stackexchange.com/a/1849294

Reply:
I removed the comment.